dapp/src/controllers/addressBookCtrl.js

```diff
@@ -0,0 +1,156 @@
+(
+  function () {
+    angular
+    .module("multiSigWeb")
+    .controller("addressBookCtrl", function ($scope, $uibModal) {
+
+      $scope.addressBook = JSON.parse(localStorage.getItem('addressBook') || '{}');
+
+      $scope.showEmptyBook = function () {
+        return Object.keys($scope.addressBook).length == 0;
+      };
+
+      /**
+       *  Open a modal and adds an item to the address book
+       */
+      $scope.addAddress = function () {
+
+        $uibModal.open({
+          templateUrl: 'partials/modals/addAddressToBook.html',
+          size: 'md',
+          scope: $scope,
+          resolve: {
+            addressBook: $scope.addressBook
+          },
+          controller: function ($scope, $uibModalInstance, addressBook, Utils, Web3Service, Wallet) {
+            $scope.book = {
+              name: '',
+              address: '',
+              type: ''
+            }
+
+            var types = {
+              multisig: 'Multisig',
+              token: 'Token',
+              eoa: 'EOA',
+              contract: 'Contract'
+            }
+
+            /**
+             * Adds and item to address book
+             */
+            $scope.addToBook = function (bookItem) {
+              // Save new address into localstorage
+              addressBook[bookItem.address] = bookItem;
+              localStorage.setItem("addressBook", JSON.stringify(addressBook));
+            };
+
+            /**
+             * Execute `addAddress` flow
+             */
+            $scope.ok = function () {
+              // Check whether it is a contract, a token or a EOA
+              var checksumAddress;
+              try {
+                checksumAddress = Web3Service.toChecksumAddress($scope.book.address);
+              } catch (error) {
+                Utils.dangerAlert(error);
+              }
+
+              if (!Web3Service.web3.isAddress($scope.book.address)) {
+                // Show alert
+                Utils.dangerAlert("Invalid Ethereum Address");
+                return;
+              }
+
+              $scope.book.address = checksumAddress;
+              
+              Web3Service.web3.eth.getCode($scope.book.address, function (e, code){
+                var hexBytecode = code.slice(-992);
+                if (code.length > 100 && Wallet.json.multiSigDailyLimit.binHex.slice(-992) == hexBytecode){
```

Comment:
with this check, we can only know for multisigs with the same compiler version and parameters, it's nice, but will not cover many instances

Reply:
Yes, that's right, good catch! I will improve it to cover more multisig instances, otherwise those multisigs would get detected as generic contracts meantime :)

Reply:
I updated that piece of code with your suggestions @denisgranha, thank you :)